### DISCUSSION TITLE
post/maintainable-winforms

### DISCUSSION THREAD

**Discussion by giscus** [markheath/soundcode-blog-comments#70] [Announcements]
# post/maintainable-winforms

Mark Heath's Development Blog

https://markheath.net/post/maintainable-winforms

**Comment by kiquenet** [comment]
how-to hig performance for Windows Forms?

**Comment by Crazyloon** [comment]
Thanks for this article Mark. I was working on an open source project someone started for fun, and users started asking for more and more features. We were originally doing everything in the form's code-behind with a few services sprinkled in.

Virtually nothing was unit tested, and we were spending a lot of time manually testing things, finding bugs, and having to redo all that testing each time we updated something minor.

I spent a week refactoring to MVP with EventAggregator and, while there is now a LOT of boilerplate code to write for each new feature, the maintenance is a breeze in comparison!

In your other article, you created the EventAggregator explicitly as a singleton, which made it a bit harder to unit test some of our scenarios, so for now, we've refactored that to not explicitly be a singleton, though it's still used as one. If we introduce DI later it will be a bit cleaner to enforce that, but for now everything is smooth sailing!

**Reply by markheath** [reply]
Glad you found it helpful! This is quite an old article, and I think if I was writing it now, I would avoid the singleton with DI. Might also consider something like Mediatr
